### ISSUE TITLE
Proposal: type guard function for the NonNullable<T> type

### ISSUE THREAD

**Issue by yokuze** [silvermine/toolbox#22]
### Scenario

We often need to filter `undefined` or `null` values from arrays. The Array#filter method is typed so that if the callback function passed to the `filter` method is a type guard, `filter`'s return type is the same as what the type guard asserts:

```typescript
const arr: number | undefined = [ 1, undefined, 2 ];

const onlyNumbers = arr.filter((n): n is number {
   return !isUndefined(n);
});
// onlyNumbers is automatically typed as `number[]`
```

Instead of writing these type guard functions ad-hoc, it would be convenient if we had a function that asserts that a variable is `NonNullable<T>`. Then, we could just:

```typescript
const onlyNumbers = arr.filter(isNotUndefinedOrNull);
```

### Naming

Possible names for the type guard function: `isNotUndefinedOrNull`, `isNonNullable`

**Comment by jthomerson** [comment]
Drive-by input: While `isNotUndefinedOrNull` is definitely a bit of a mouthful, it's a ton more intuitive to me than `isNonNullable` ... I'd see `isNonNullable` and never think it would apply to `undefined` intuitively.

**Comment by yokuze** [comment]
I think the same. `NonNullable` seems like a TypeScript-specific name (and not a clear one at that)... unless it's some comp sci concept I'm just not aware of.

**Comment by yokuze** [comment]
Addressed with #33
